Include/internal/pycore_object_alloc.h

```diff
@@ -0,0 +1,71 @@
+#ifndef Py_INTERNAL_OBJECT_ALLOC_H
+#define Py_INTERNAL_OBJECT_ALLOC_H
+
+#include "pycore_object.h"      // _PyType_HasFeature()
+#include "pycore_pystate.h"     // _PyThreadState_GET()
+#include "pycore_tstate.h"      // _PyThreadStateImpl
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#ifndef Py_BUILD_CORE
+#  error "this header requires Py_BUILD_CORE define"
+#endif
+
+#ifdef Py_GIL_DISABLED
+static inline mi_heap_t *
+_PyObject_GetAllocationHeap(_PyThreadStateImpl *tstate, PyTypeObject *tp)
+{
+    struct _mimalloc_thread_state *m = &tstate->mimalloc;
+    if (_PyType_HasFeature(tp, Py_TPFLAGS_PREHEADER)) {
+        return &m->heaps[_Py_MIMALLOC_HEAP_GC_PRE];
+    }
+    else if (_PyType_IS_GC(tp)) {
+        return &m->heaps[_Py_MIMALLOC_HEAP_GC];
+    }
+    else {
+        return &m->heaps[_Py_MIMALLOC_HEAP_OBJECT];
+    }
+}
+#endif
+
+// Sets the heap used for PyObject_Malloc(), PyObject_Realloc(), etc. calls in
+// Py_GIL_DISABLED builds. We use different heaps depending on if the object
+// supports GC and if it has a pre-header. We smuggle the choice of heap
+// through the _mimalloc_thread_state. In the default build, this simply
+// calls PyObject_Malloc().
+static inline void *
+_PyObject_MallocWithType(PyTypeObject *tp, size_t size)
+{
+#ifdef Py_GIL_DISABLED
+    _PyThreadStateImpl *tstate = (_PyThreadStateImpl *)_PyThreadState_GET();
+    struct _mimalloc_thread_state *m = &tstate->mimalloc;
+    m->current_object_heap = _PyObject_GetAllocationHeap(tstate, tp);
+#endif
+    void *mem = PyObject_Malloc(size);
+#ifdef Py_GIL_DISABLED
+    m->current_object_heap = &m->heaps[_Py_MIMALLOC_HEAP_OBJECT];
```

Comment:
You can call `PyObject_Malloc()` directly -- that doesn't go through this API. Only calls that might allocate GC objects need to go through this. So we have to restore `current_object_heap` to the non-GC object heap.

You could structure this a bit differently and have `PyObject_Malloc()` also go through this sort of thing, but I think that would end up being more complicated.

Reply:
But if you call `PyObject_Malloc` directly aren't things now broken with the pre-header?  It seems that `Py_TPFLAGS_MANAGED_DICT` and `Py_TPFLAGS_MANAGED_WEAKREF` are both documented as being available to extension objects which would be the primary callers of `PyObject_Malloc`.  Maybe we at least need to document that if you're using these flags you need to call `PyType_GenericAlloc` in `Doc/c-api/typeobj.rst`.

I suppose before this you'd be probably broken if you called `PyObject_Malloc`, but it would have crashed and burned pretty obviously, and you could figure out to call `_PyType_PreHeaderSize`.  Now it seems critical that you call `PyType_GenericAlloc` or `_PyObject_MallocWithType` or you'll get really subtle crashes in multi-threaded scenarios.

Reply:
You can call `PyObject_Malloc()` for non-GC objects, but not for GC-objects. GC objects have had "pre-headers" (`PyGC_Head`) before `Py_TPFLAGS_MANAGED_DICT` was introduced. 

These restrictions have been in place since Python 2.2a3:
https://github.com/python/cpython/blob/72abb8c5d487ead9eb115fec8132ccef5ba189e5/Misc/HISTORY#L24391-L24398

Reply:
Ahh, true, I hadn't considered that the pre-header implies GC, that makes sense!